components/search/custom-instant-search.tsx

```diff
@@ -0,0 +1,201 @@
+import { useCallback, useState } from "react";
+
+import CustomHits from "~/components/search/custom-hits";
+import CustomRefinementList from "~/components/search/custom-refinement-list";
+import CustomSearchBox from "~/components/search/custom-search-box";
+import { RefinementModal } from "~/components/search/refinement-modal";
+import { getQueryParams } from "~/lib/string-utils";
+
+import { useRouter } from "next/router";
+import {
+  Configure,
+  connectRefinementList,
+  SearchState,
+} from "react-instantsearch-core";
+import {
+  InstantSearch,
+  InstantSearchProps,
+  RefinementItem,
+} from "react-instantsearch-dom";
+import { debounce } from "ts-debounce";
+
+type FilterSetting = {
+  field: string;
+  title: string;
+};
+
+interface CustomInstantSearchProps extends InstantSearchProps {
+  itemName: string;
+  filterSettings?: FilterSetting[];
+  useFilterModal?: boolean;
```

Comment:
```suggestion
  withFilterModal?: boolean;
```

`useFilterModal` -> `withFilterModal` - since functions starting with `use` is a common convention for React hooks, we don't want to confuse this prop with a hook. 😅